pkg/streamingpromql/optimize/ast/propagate_matchers.go

```diff
@@ -0,0 +1,312 @@
+	}
+	groupingSet := newStringSet(e.Grouping)
+	for _, vs := range vss {
+		updateVectorSelectorWithAggregationInfo(vs, groupingSet, include)
+	}
+	newMatchers := mapper.getMatchersToPropagate(labelMatchers, groupingSet, include)
+	return vss, newMatchers
+}
+
+func updateVectorSelectorWithAggregationInfo(vs *enrichedVectorSelector, groupingSet stringSet, include bool) {
+	if vs.containsAgg {
+		switch {
+		case vs.include && include:
+			vs.labelsSet = vs.labelsSet.Intersection(groupingSet)
+		case !vs.include && !include:
+			vs.labelsSet = vs.labelsSet.Union(groupingSet)
+		case vs.include && !include:
+			vs.labelsSet = vs.labelsSet.Difference(groupingSet)
+		case !vs.include && include:
+			vs.labelsSet = groupingSet.Difference(vs.labelsSet)
+			vs.include = true
+		}
+		return
+	}
+	vs.labelsSet = groupingSet
+	vs.include = include
+	vs.containsAgg = true
+}
+
+func VectorSelectorArgumentIndex(funcName string) (int, error) {
+	// If the function is eligible for matcher propagation, return the index of the argument
+	// that contains the vector selector (may be embedded in other kinds of expressions).
+	// Otherwise, return -1.
+	switch funcName {
+	// Simple
+	case "absent", "changes", "resets":
+		return 0, nil
+	// Mathematical
+	case "abs", "sgn", "sqrt", "exp", "deriv", "ln", "log2", "log10", "ceil", "floor", "round", "clamp", "clamp_min", "clamp_max":
+		return 0, nil
+	// More complicated mathematical functions
+	case "double_exponential_smoothing", "predict_linear":
+		return 0, nil
+	// Trigonometric
+	case "acos", "acosh", "asin", "asinh", "atan", "atanh", "cos", "cosh", "sin", "sinh", "tan", "tanh", "deg", "rad":
+		return 0, nil
+	// Differences between samples (using matrix selectors)
+	case "rate", "delta", "increase", "idelta", "irate":
+		return 0, nil
+	// Histogram-related
+	case "histogram_avg", "histogram_count", "histogram_sum", "histogram_stddev", "histogram_stdvar":
+		return 0, nil
+	case "histogram_quantile":
+		return 1, nil
+	case "histogram_fraction":
+		return 2, nil
+	// Aggregation over time
+	case "avg_over_time", "min_over_time", "max_over_time", "sum_over_time", "count_over_time", "stddev_over_time", "stdvar_over_time", "last_over_time", "present_over_time", "absent_over_time", "mad_over_time", "ts_of_min_over_time", "ts_of_max_over_time", "ts_of_last_over_time":
+		return 0, nil
+	case "quantile_over_time":
+		return 1, nil
+	// Time
+	case "minute", "hour", "day_of_week", "day_of_month", "day_of_year", "days_in_month", "month", "year", "timestamp":
+		return 0, nil
+	// No vector/matrix selectors
+	case "pi", "time", "vector":
+		return -1, nil
+	// Explicitly not supported because it's not valid to propagate matchers across these functions.
+	case "scalar":
+		return -1, nil
+	// Explicitly not supported because we want to avoid unexpected interactions with labels or ordering.
+	case "label_join", "label_replace", "info", "sort", "sort_desc", "sort_by_label", "sort_by_label_desc":
+		return -1, nil
```

Comment:
Could you please add some tests for these cases to make sure we don't inadvertently break this in the future?